Feature/config migration[WIP]

pksjce opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->
#### Migrations
- [x] loaders -> rules
- [x] pre/postLoaders -> rules with `enforce: 'pre/post'`
- [ ] solve for `-loaders` suffix.
- [ ] remove `json-loader`
- [ ] add `sourceMap: true` to `UglifyJsPlugin`
- [ ] change `ExtractTextPlugin` syntax
- [ ] remove `OccurendeOrderPlugin`
- [ ] add `LoaderOptionsPlugin` with `minimize:true` ,`context` and `debug`

#### NFR

- [x] Tests should pass
- [ ] Add yarn support
- [x] Add node 4 support
**What kind of change does this PR introduce?**
Adds a codemod to change webpack1 loader syntax to webpack2 syntax

**Did you add tests for your changes?**
Yes. 

**If relevant, did you update the documentation?**
Not yet, once we have some more and attach it to the CLI then can document

**Summary**
It would be really great for the project if webpack1->2 migration can be done through codemods. As described in https://github.com/webpack/webpack-cli/issues/6

**Does this PR introduce a breaking change?**
No

**Other information**

pksjce commented:
The codemod is doing it's job, but tests are failing due to formatting problems in the replaced and pretty printed source. :( Still trying to figure that out.

ev1stensberg commented on lib/transformations/loaders/loaders.js:
```diff
@@ -0,0 +1,83 @@
+
+	const loadersToRules = () => ast
+    .find(j.Identifier)
+    .filter(p => p.value.name === 'loaders' && safeTraverse(p, ['parent', 'parent', 'parent', 'node', 'key', 'name']) === 'module')
```
If there's an error in your code, here is probably where it is. Adds whitespace for some reason.

ev1stensberg commented on lib/transformations/loaders/loaders.js:
```diff
@@ -0,0 +1,83 @@
+	const loaderToUse = () => ast
+    .find(j.Identifier)
+    .filter(p => (p.value.name === 'loaders' || p.value.name === 'loader') && safeTraverse(p, ['parent', 'parent', 'parent', 'parent', 'node', 'key', 'name']) === 'rules')
+    .forEach(p => p.value.name = 'use')
```
If there's an error in your code, here is probably where it is. Adds whitespace for some reason.

ev1stensberg commented:
This bug seems a bit wierd to debug at first glance. Try to create multiple test files with different indentations and log them to find out where the `loc` prop fails. (See line 55). Do `.forEach(p => console.log(p.name)`

pksjce commented:
So fixed multiple things. Seemed like a rabbithole. But got everything to pass at the end of it.
1. Filed this [issue](https://github.com/benjamn/recast/pull/353). This identifies the bug with the newline but has only a temp fix. Need more input from the maintainers there.
2. Added [my branch](https://github.com/kalcifer/recast/tree/bug/allowbreak) as the package source for `recast`. But this means `yarn` will not work on travis. So moved yarn.lock to backup file temporarily. Once the issue in `recast` is fixed, I will move to normalcy. Will keep this PR in WIP until then.
3. Created my own testUtil from the original and added `eslint` api to fix indentation for testing.

ev1stensberg commented:
Could you put this PR on hold until `yarn` works? Or bypass this somehow?

ev1stensberg commented:
>> So moved yarn.lock to backup file temporarily

Elaborate please 💯 

- What's the issue with node 4 support?

pksjce commented:
I hope this to be temporary till I get some clarity on previous issue.

 - [yarn does not support github links](https://github.com/yarnpkg/yarn/issues/513). PR is under review for this. 
 - Will add back node4 after some investigation.

ev1stensberg commented:
Oo, that's unfortunate. Get v.4 to work (disable yarn if needed) and we'll see!

pksjce commented:
Yep 👍  Done!

ev1stensberg commented:
Thanks 🎉

ev1stensberg commented:
Put up a issue also, so we don't forget to revert the yarn error once it's fixed.
